codegen/masm/src/codegen.rs

```diff
@@ -630,7 +703,12 @@ impl<'ast> AirVisitor<'ast> for CodeGenerator<'ast> {
                     periodic_column_to_target_el(column),
                 )?;
             }
-            Value::PublicInput(_, _) => todo!(),
+            Value::PublicInput(name, index) => {
+                self.writer
+                    .header(format!("Load public input {} pos {}", name, index));
+                let index: u64 = (*index).try_into().or(Err(CodegenError::InvalidIndex))?;
+                load_quadratic_element(&mut self.writer, self.config.public_inputs_address, index)?;
+            }
```

Comment:
Let me use another example:

```
public_inputs:
    program_hash: [4]
    stack_inputs: [4]
    stack_outputs: [20]
    overflow_addrs: [4]
```

These are the open questions:

- Is the order of the inputs guaranteed to be stable through out the pipeline so that the indexes can be computed as you specified?
    - IMO we should not rely on text order, since most of the language feels like a mathematical instead of operation description, i.e. I expect people will move rows around without realizing that this breaks the system, which is bad for maintenance and language evolution. I would do what is done for other descriptive languages, like protobuf, and add a numerical ordering to decouple it from text position
- What is the plan to determine with input is against the first and last row? This is necessary to compute the divisor of the boundary constraint. Which is what I was referring to as an issue.

Reply:
Currently, it is up to each backend to specify how public inputs are serialized/deserialized - so, to keep the order consistent, we'll need to make sure both Winterfell and MASM back-ends do this consistently. I think we have a few options here:

1. Assume that inputs are serialized/deserialized in the order in which they were declared. I agree with you that this is probably not a good option.
2. Sort the inputs by their name, and serialize in that order. This does mean that we'll need to ensure consistency across back-end by hand.
3. Explicitly describe serialization order in AirScript.

I think for the time being, we could go with option 2 (sort by name). And maybe later on we could switch to option 3.

Reply:
To figure out which boundary constraint goes against which row, we can use constraint domain contained in [ConstraintRoot](https://github.com/0xPolygonMiden/air-script/blob/next/ir/src/constraints/constraint.rs#L8) struct. btw, this is also how we can differentiate between transition and validity constraints (`EveryFrame` is a transition constraint and `EveryRow` is a validity constraint).